packages/frontend/src/pages/Playground.svelte

```diff
@@ -1,5 +1,151 @@
 <script lang="ts">
+import { conversations } from '../stores/conversations';
+import { studioClient } from '/@/utils/client';
+import {
+  isAssistantChat,
+  type ChatMessage,
+  type UserChat,
+  isPendingChat,
+  isUserChat,
+  type AssistantChat,
+} from '@shared/src/models/IPlaygroundMessage';
+import NavPage from '../lib/NavPage.svelte';
+import { playgrounds } from '../stores/playgrounds-v2';
+import { catalog } from '../stores/catalog';
+import Button from '../lib/button/Button.svelte';
+import { afterUpdate } from 'svelte';
+
 export let playgroundId: string;
+let prompt: string;
+let sendEnabled = false;
+let scrollable: Element;
+let lastIsUserMessage = false;
+let errorMsg = '';
+
+$: conversation = $conversations.find(conversation => conversation.id === playgroundId);
+$: playground = $playgrounds.find(playground => playground.id === playgroundId);
+$: model = $catalog.models.find(model => model.id === playground?.modelId);
+$: {
+  if (conversation?.messages.length) {
+    const latest = conversation.messages[conversation.messages.length - 1];
+    if (isAssistantChat(latest) && !isPendingChat(latest)) {
+      sendEnabled = true;
+    }
+    lastIsUserMessage = isUserChat(latest);
+  } else {
+    sendEnabled = true;
+  }
+}
+
+const roleNames = {
+  system: 'System',
+  user: 'User',
+  assistant: 'Assistant',
+};
+
+function getMessageParagraphs(message: ChatMessage): string[] {
+  if (isAssistantChat(message)) {
+    if (!isPendingChat(message)) {
+      return message.content?.split('\n') ?? [];
+    } else {
+      return message.choices
+        .map(choice => choice.content)
+        .join('')
+        .split('\n');
+    }
+  } else if (isUserChat(message)) {
+    const msg = message as UserChat;
+    return msg.content?.split('\n') ?? [];
+  }
+  return [];
+}
+
+function askPlayground() {
+  studioClient
+    .submitPlaygroundMessage(playgroundId, prompt)
+    .then(() => {
+      errorMsg = '';
+      sendEnabled = false;
+      prompt = '';
+    })
+    .catch((err: unknown) => {
+      errorMsg = String(err);
+      sendEnabled = true;
+    });
+}
+
+afterUpdate(() => {
+  if (!conversation?.messages.length) {
+    return;
+  }
+  const latest = conversation.messages[conversation.messages.length - 1];
+  if (isUserChat(latest) || (isAssistantChat(latest) && isPendingChat(latest))) {
+    scrollToBottom(scrollable);
+  }
+});
+
+async function scrollToBottom(element: Element) {
+  element.scroll?.({ top: element.scrollHeight, behavior: 'smooth' });
+}
+
+function elapsedTime(msg: AssistantChat): string {
```

Comment:
The way you compute the elapsed time is leading to incoherence when the inference server is stopped/restart and we return to the page.

https://github.com/projectatomic/ai-studio/assets/42176370/3ef933ee-2a60-4c13-8b62-a07f3a2bfcc4

IMO we should remove the code about the elapsed time as this PR is big enough. And make follow up PR

Reply:
The elapsed time should be computed by the backend, and not by the frontend

Reply:
That could be a follow up PR